tsl/src/nodes/decompress_chunk/decompress_chunk.c

```diff
@@ -335,57 +327,105 @@ cost_decompress_chunk(Path *path, Path *compressed_path)
+cost_batch_sorted_merge(PlannerInfo *root, CompressionInfo *compression_info,
+						DecompressChunkPath *dcpath, Path *compressed_path)
 {
 	Path sort_path; /* dummy for result of cost_sort */
 
+	/*
+	 * Don't disable the compressed batch sorted merge plan with the enable_sort
+	 * GUC. We have a separate GUC for it, and this way you can try to force the
+	 * batch sorted merge plan by disabling sort.
+	 */
+	const bool old_enable_sort = enable_sort;
+	enable_sort = true;
 	cost_sort(&sort_path,
 			  root,
 			  dcpath->compressed_pathkeys,
-			  child_path->total_cost,
-			  child_path->rows,
-			  child_path->pathtarget->width,
+			  compressed_path->total_cost,
+			  compressed_path->rows,
+			  compressed_path->pathtarget->width,
 			  0.0,
 			  work_mem,
 			  -1);
+	enable_sort = old_enable_sort;
 
-	/* startup_cost is cost before fetching first tuple */
-	dcpath->custom_path.path.startup_cost = sort_path.total_cost;
+	/*
+	 * In compressed batch sorted merge, for each distinct segmentby value we
+	 * have to keep the corresponding latest batch open. Estimate the number of
+	 * these batches to account for it in the cost model.
+	 */
+	List *segmentby_groupexprs = NIL;
+	for (int segmentby_attno = bms_next_member(compression_info->chunk_segmentby_attnos, -1);
+		 segmentby_attno > 0;
+		 segmentby_attno =
+			 bms_next_member(compression_info->chunk_segmentby_attnos, segmentby_attno))
+	{
+		Var *var = palloc(sizeof(Var));
+		*var = (Var){ .xpr.type = T_Var,
+					  .varno = compression_info->compressed_rel->relid,
+					  .varattno = segmentby_attno };
+		segmentby_groupexprs = lappend(segmentby_groupexprs, var);
+	}
+
+	const double simultaneously_open_baches =
+		estimate_num_groups(root, segmentby_groupexprs, dcpath->custom_path.path.rows, NULL, NULL);
 
 	/*
-	 * The cost model for the normal chunk decompression produces the following total
-	 * costs.
-	 *
-	 * Segments  Total costs
-	 * 10         711.84
-	 * 50        4060.91
-	 * 100       8588.32
-	 * 10000   119281.84
-	 *
-	 * The cost model of the regular decompression is roughly linear. Opening multiple batches in
-	 * parallel needs resources and merging a high amount of batches becomes inefficient at some
-	 * point. So, we use a quadratic cost model here to have higher costs than the normal
-	 * decompression when more than ~100 batches are used. We set
-	 * DECOMPRESS_CHUNK_HEAP_MERGE_CPU_TUPLE_COST to 0.8 to become most costly as soon as we have to
-	 * process more than 120 batches.
-	 *
-	 * Note: To behave similarly to the cost model of the regular decompression path, this cost
-	 * model does not consider the number of tuples.
+	 * Getting one uncompressed row involves replacing the top row in the heap,
+	 * which costs O(log(heap size)).
+	 * The constant multiplier is found empirically by benchmarking queries returning
+	 * 1 - 1e9 tuples with segmentby cardinality 1 to 1e4, and adjusting the
+	 * cost so that the fastest plan is used.
```

Comment:
The multiplier is chosen by this very scientific benchmark: https://gist.github.com/akuzm/a0c3c68a0ecf7e14a31f88954f3e3eec